Add constants for HTTP status codes

annevk opened the pull request:
#### 7b1232f1e3b0a361fdf7d81dd185c4943fe9267d
<pre>
Add constants for HTTP status codes
<a href="https://bugs.webkit.org/show_bug.cgi?id=268224">https://bugs.webkit.org/show_bug.cgi?id=268224</a>

Reviewed by Simon Fraser and Youenn Fablet.

Implement an idea from Simon Fraser to collect status codes in a
centralized file with the aim to improve readability. As a side effect
it might make it easier to find related code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::BlobLoader::didReceiveResponse):
* Source/WebCore/Modules/fetch/FetchResponse.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::didReceiveResponse):
(WebCore::FileReaderLoader::httpStatusCodeToErrorCode):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didReceiveBuffer):
(WebCore::SubresourceLoader::responseHasHTTPStatusCodeError const):
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::responseIsValid const):
* Source/WebCore/platform/network/HTTPStatusCodes.h: Added.
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::shouldUseGet):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::synthesizedResponseForRange):
(WebCore::RangeResponseGenerator::willSynthesizeRangeResponses):
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::willSendRequest):
(WebCore::ResourceHandle::tryHandlePasswordBasedAuthentication):
(WebCore::ResourceHandle::receivedCredential):
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::validateResponse):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didReceiveInformationalResponse):
(WebKit::NetworkResourceLoader::didReceiveResponse):
(WebKit::NetworkResourceLoader::didFinishLoading):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.cpp:
(WebKit::ServiceWorkerNavigationPreloader::didReceiveResponse):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerSoftUpdateLoader.cpp:
(WebKit::ServiceWorkerSoftUpdateLoader::didReceiveResponse):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::store):
* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoad.cpp:
(WebKit::NetworkCache::SpeculativeLoad::didReceiveResponse):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::tryPasswordBasedAuthentication):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
(WebKit::CompletionHandler&lt;void):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(WebKit::PopUpSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::completeInternal):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge):
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::PDFPluginStreamLoaderClient::didReceiveResponse):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::shouldSkipDecidePolicyForResponse const):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::SocketStreamHandleImpl::readStreamCallback):
(WebCore::SocketStreamHandleImpl::writeStreamCallback):

Canonical link: <a href="https://commits.webkit.org/273747@main">https://commits.webkit.org/273747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8c0ad91b597f789cb8f53b698643e35578e1af6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36547 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39209 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32778 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17959 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12566 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31396 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37107 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13054 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32330 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11424 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11450 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40454 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33119 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32929 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37376 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11689 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35481 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13379 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8284 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12119 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12575 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/HTTPStatusCodes.h:
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+namespace WTF {
```
I am not convinced that this header belongs in wtf/. I would have put it under WebCore/platform/network.

annevk replied:
It touches files outside of WebCore. Is that okay? Happy to move it.

cdumez replied:
The other files are in WebKit so this should be fine. Being in WebCore would only be an issue if you needed it in JavaScriptCore as far as I know

cdumez replied:
You just need to make sure you mark the header as private in Xcode so it gets exported.

annevk replied:
Alright, sounds good. Thanks! Will move it tomorrow and maybe find a couple more files to update while I'm at it.

youennf commented on Source/WebCore/platform/network/HTTPStatusCodes.h:
```diff
@@ -0,0 +1,64 @@
+constexpr auto httpStatusTemporaryRedirect = 307;
+constexpr auto httpStatusPermanentRedirect = 308;
+
+constexpr auto httpStatusBadRequest = 400;
```
I am surprised we do not have 401.

youennf commented on Source/WebCore/platform/network/HTTPStatusCodes.h:
```diff
@@ -0,0 +1,64 @@
+
+#pragma once
+
+namespace WTF {
```
s/WTF/WebCore.

donny-dont replied:
If this is actually WTF then it should live in WTF. I would prolly argue that this is something that should ideally be in PAL.

Also I've seen a lot of nested namespaces in the updated layout code. I think this would be a good case here for using a nested namespace so its `HttpStatus::ok`  instead of `httpStatusOk`. Would definitely help with auto complete.

annevk replied:
As identifiers cannot start with numbers I could not figure out a nice way to do it. `HTTPStatus::OK200` doesn't read as well as `httpStatus200OK`. And `HTTPStatus::_200OK` seems ungainly.

youennf commented on Source/WebCore/platform/network/HTTPStatusCodes.h:
```diff
@@ -0,0 +1,64 @@
+
+// https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
+
+constexpr auto httpStatusOk = 200;
```
s/Ok/OK ?

I would also have thought that http would be upper cased, not sure what the coding style is saying.

annevk replied:
Should be lowercase per https://webkit.org/code-style-guidelines/#names-basic. And yeah, I guess it should be OK.

youennf commented on Source/WebCore/platform/network/HTTPStatusCodes.h:
```diff
@@ -0,0 +1,64 @@
+constexpr auto httpStatusPermanentRedirect = 308;
+
+constexpr auto httpStatusBadRequest = 400;
+constexpr auto httpStatusForbidden = 403;
```
Ditto for 404.

youennf commented on Source/WebCore/loader/SubresourceLoader.cpp:
```diff
@@ -416,7 +417,7 @@ void SubresourceLoader::didReceiveResponse(const ResourceResponse& response, Com
     // cases of too many redirects from CFNetwork (<rdar://problem/30610988>).
 #if !PLATFORM(COCOA)
-    ASSERT(response.httpStatusCode() < 300 || response.httpStatusCode() >= 400 || response.httpStatusCode() == 304 || response.httpHeaderField(HTTPHeaderName::Location).isEmpty() || response.type() == ResourceResponse::Type::Opaqueredirect);
+    ASSERT(response.httpStatusCode() < 300 || response.httpStatusCode() >= 400 || response.httpStatusCode() == httpStatusNotModified || response.httpHeaderField(HTTPHeaderName::Location).isEmpty() || response.type() == ResourceResponse::Type::Opaqueredirect);
```
Why not replacing 400, we are doing it below?
Or if we do not do it here, let's not do it below.

donny-dont commented on Source/WebDriver/socket/HTTPServerSocket.cpp:
```diff
@@ -117,7 +118,7 @@ String HTTPRequestHandler::packHTTPMessage(HTTPRequestHandler::Response&& respon
     const char* EOL = "\r\n";
 
-    builder.append("HTTP/1.0 ", response.statusCode, ' ', response.statusCode == 200 ? "OK" : "ERROR", EOL);
+    builder.append("HTTP/1.0 ", response.statusCode, ' ', response.statusCode == httpStatusOk ? "OK" : "ERROR", EOL);
```
So this failed the build because this patch introduced a dependency on WebCore for platforms using the Socket based inspector, aka Windows and PlayStation. Given that its just one place that this is used I would lean towards just reverting back to the previous version instead of adding a build dependency.

donny-dont reviewed (commented):
Had some comments along with how to fix the Windows and PlayStation build with this patch.

aproskuryakov commented on Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
```diff
@@ -308,7 +309,7 @@ FetchBodyOwner::BlobLoader::BlobLoader(FetchBodyOwner& owner)
 void FetchBodyOwner::BlobLoader::didReceiveResponse(const ResourceResponse& response)
 {
-    if (response.httpStatusCode() != 200)
+    if (response.httpStatusCode() != httpStatusOk)
```
FWIW, I personally would find this less readable - I know what status 200 is, but not what `Ok` is.

cdumez replied:
Maybe we should just use the status code in the constant name? e.g. "httpStatus200" or "httpStatusCode200"?

annevk replied:
https://www.youtube.com/watch?v=vqgSO8_cRio

annevk committed:
Add constants for HTTP status codes
https://bugs.webkit.org/show_bug.cgi?id=268224

Reviewed by Simon Fraser and Youenn Fablet.

Implement an idea from Simon Fraser to collect status codes in a
centralized file with the aim to improve readability. As a side effect
it might make it easier to find related code.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchBodyOwner.cpp:
(WebCore::FetchBodyOwner::BlobLoader::didReceiveResponse):
* Source/WebCore/Modules/fetch/FetchResponse.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/fileapi/FileReaderLoader.cpp:
(WebCore::FileReaderLoader::didReceiveResponse):
(WebCore::FileReaderLoader::httpStatusCodeToErrorCode):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didReceiveResponse):
(WebCore::SubresourceLoader::didReceiveBuffer):
(WebCore::SubresourceLoader::responseHasHTTPStatusCodeError const):
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::responseIsValid const):
* Source/WebCore/platform/network/HTTPStatusCodes.h: Added.
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::shouldUseGet):
* Source/WebCore/platform/network/cocoa/RangeResponseGenerator.mm:
(WebCore::synthesizedResponseForRange):
(WebCore::RangeResponseGenerator::willSynthesizeRangeResponses):
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::willSendRequest):
(WebCore::ResourceHandle::tryHandlePasswordBasedAuthentication):
(WebCore::ResourceHandle::receivedCredential):
* Source/WebKit/NetworkProcess/NetworkLoadChecker.cpp:
(WebKit::NetworkLoadChecker::validateResponse):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didReceiveInformationalResponse):
(WebKit::NetworkResourceLoader::didReceiveResponse):
(WebKit::NetworkResourceLoader::didFinishLoading):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerNavigationPreloader.cpp:
(WebKit::ServiceWorkerNavigationPreloader::didReceiveResponse):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerSoftUpdateLoader.cpp:
(WebKit::ServiceWorkerSoftUpdateLoader::didReceiveResponse):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::store):
* Source/WebKit/NetworkProcess/cache/NetworkCacheSpeculativeLoad.cpp:
(WebKit::NetworkCache::SpeculativeLoad::didReceiveResponse):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::tryPasswordBasedAuthentication):
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
(WebKit::CompletionHandler<void):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/PopUpSOAuthorizationSession.mm:
(WebKit::PopUpSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/RedirectSOAuthorizationSession.mm:
(WebKit::RedirectSOAuthorizationSession::completeInternal):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/SubFrameSOAuthorizationSession.mm:
(WebKit::SubFrameSOAuthorizationSession::completeInternal):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebRequestCocoa.mm:
(WebKit::WebExtensionContextProxy::resourceLoadDidReceiveChallenge):
* Source/WebKit/WebProcess/Plugins/PDF/PDFIncrementalLoader.mm:
(WebKit::PDFPluginStreamLoaderClient::didReceiveResponse):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::shouldSkipDecidePolicyForResponse const):
* Source/WebKitLegacy/WebCoreSupport/SocketStreamHandleImplCFNet.cpp:
(WebCore::SocketStreamHandleImpl::readStreamCallback):
(WebCore::SocketStreamHandleImpl::writeStreamCallback):

Canonical link: https://commits.webkit.org/273747@main
